File: Greedy_algorithm/907/interval.cpp
#include <iostream>
#include <algorithm>
using namespace std;

const int N = 1e5 + 10;
int n, l, r, s, t, res;
struct Interval{
    int l, r;
    bool operator < (const struct Interval &interval) {
        return l < interval.l;
    }
} e[N];

int main() {
    scanf("%d%d%d", &s, &t, &n);
    for (int i = 0; i < n; i++) {
        scanf("%d%d", &l, &r);
        e[i] = {l, r};
    }
    sort(e, e + n);
    for (int i = 0, j = 0; i < n; i++) {
        r = -1e9 - 10;
        while (j < n && e[j].l <= s) {
            r = max(r, e[j++].r);
        }
        if (r < s) break;
        res++;
        if (r >= t) break;
        s = r;
        i = j - 1;
    }
    if (r < t) res = -1;
    printf("%d", res);
    return 0;
}

File: Greedy_algorithm/907/interval.py
if __name__ == "__main__":
    s, t = map(int, input().split())
    n = int(input())
    e = []
    for _ in range(n):
        e.append(tuple(map(int, input().split())))
    e.sort()
    res = 0
    i = j = 0
    while i < n:
        r = -0x3f3f3f3f
        while j < n and e[j][0] <= s:
            r = max(r, e[j][1])
            j += 1
        res += 1
        if r < s or r >= t:
            break
        s = r
        i = j
    if r < t:
        res = -1
    print(res)
